src/utils.cpp

```diff
@@ -246,6 +248,107 @@ bool UnsatCoreReducer::reduce_assump_unsatcore(const Term &formula,
   return true;
 }
 
+bool UnsatCoreReducer::linear_reduce_assump_unsatcore(
+                              const smt::Term &formula,
+                              const smt::TermVec &assump,
+                              smt::TermVec &out_red,
+                              smt::TermVec *out_rem,
+                              unsigned iter)
+{
+  TermVec bool_assump, tmp_assump;
+  UnorderedTermMap to_ext_assump;
+  TermVec cand_res;
+  for (const auto & a : assump) {
+    Term t = to_reducer_.transfer_term(a);
+    cand_res.push_back(t);
+    to_ext_assump[t] = a;
+  }
+
+  reducer_->push();
+  reducer_->assert_formula(to_reducer_.transfer_term(formula));
+
+  // exit if the formula is unsat without assumptions.
+  Result r = reducer_->check_sat();
+  if (r.is_unsat()) {
+    reducer_->pop();
+    return true;
+  }
+
+  UnorderedTermMap label_to_cand_;
+  for (const auto & a : cand_res) {
+    Term l = label(a);
+    reducer_->assert_formula(reducer_->make_term(Implies, l, a));
+    bool_assump.push_back(l);
+    label_to_cand_.emplace(l, a);
+  }
+
+  unsigned cur_iter = 0;
+  unsigned assump_pos = 0;
+  r = reducer_->check_sat_assuming(bool_assump);
+  if (r.is_sat()) {
+    reducer_->pop();
+    return false;
+  }
+  assert(r.is_unsat());
+  TermVec bool_assump_query;
+
+  while (cur_iter <= iter && assump_pos < bool_assump.size()) {
+    cur_iter = iter > 0 ? cur_iter+1 : cur_iter;
+    bool_assump_query = bool_assump;
+    bool_assump_query.erase(bool_assump_query.begin()+assump_pos);
+    r = reducer_->check_sat_assuming(bool_assump_query);
+    if (r.is_sat()) {
+      // we cannot remove this assumption, then try next one
+      ++ assump_pos;
+    } else {
+      // we can remove this assumption
```

Comment:
where do we remove this assumption? `assump_pos`

Reply:
So I think it will be removed below because it by definition is not in the unsat core (since it wasn't assumed in this iteration). Although, one minor thing to note is that `vector::erase` is not very efficient. It might even be better to clear the vector and just add things back into it.

Reply:
Right.